Drop TextRun::characters8() / characters16()

cdumez opened the pull request:
#### f74abf65a72c20f20ac736065a7467643de06652
<pre>
Drop TextRun::characters8() / characters16()
<a href="https://bugs.webkit.org/show_bug.cgi?id=273199">https://bugs.webkit.org/show_bug.cgi?id=273199</a>

Reviewed by Darin Adler.

Drop TextRun::characters8() / characters16() in favor of span versions.

* Source/WebCore/layout/formattingContexts/inline/text/TextUtil.cpp:
(WebCore::Layout::TextUtil::fallbackFontsForText):
* Source/WebCore/platform/graphics/ComplexTextController.cpp:
(WebCore::ComplexTextController::collectComplexTextRuns):
* Source/WebCore/platform/graphics/FontCascade.cpp:
(WebCore::computeUnderlineType):
* Source/WebCore/platform/graphics/TextRun.h:
* Source/WebCore/platform/graphics/WidthIterator.cpp:
(WebCore::WidthIterator::advance):

Canonical link: <a href="https://commits.webkit.org/277986@main">https://commits.webkit.org/277986@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/3663d6236fbb782f326338b25558a83206fef6d2

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/49138 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/28415 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/52150 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/51936 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/45232 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/51442 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/34397 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/25988 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/40162 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/50779 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/25979 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/42357 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/21275 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/23435 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/43531 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/7462 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/45363 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/44038 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/53845 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/24216 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/20421 "Found 1 new test failure: media/video-pause-immediately.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/47481 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/25494 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/42562 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/46467 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/26308 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/7042 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/25213 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/platform/graphics/SurrogatePairAwareTextIterator.h:
```diff
@@ -29,8 +29,8 @@ class SurrogatePairAwareTextIterator {
     // 'endIndex' denotes the maximum length of the UChar array, which might exceed 'lastIndex'.
-    SurrogatePairAwareTextIterator(const UChar* characters, unsigned currentIndex, unsigned lastIndex, unsigned endIndex)
-        : m_characters(characters)
+    SurrogatePairAwareTextIterator(std::span<const UChar> characters, unsigned currentIndex, unsigned lastIndex, unsigned endIndex)
```
We should use `characters.size() + currentIndex` as `endIndex`, right?

cdumez replied:
Seems like this would be a behavior change?

darinadler replied:
I don’t think so, no. I looked at all the call sites. Can you find one where it’s not correct?

cdumez replied:
This one in WidthIterator?
```
SurrogatePairAwareTextIterator textIterator(m_run->subspan16(m_currentCharacterIndex), m_currentCharacterIndex, offset, length);
```

As far as I can tell, `length == m_run->span16().size()` but `m_currentCharacterIndex` is not necessarily 0?

cdumez replied:
Oh never mind, I'm tired. `m_run->subspan16(m_currentCharacterIndex).size() == m_run->span16().size() - m_currentCharacterIndex`. I'll make the change.

cdumez replied:
But then, shouldn't we use `endIndex = characters.size() + currentIndex` instead of `endIndex = characters.size() - currentIndex`?

darinadler replied:
Yes, I had it wrong. I edited it above to the correct suggestion.

darinadler commented on Source/WebCore/platform/graphics/Latin1TextIterator.h:
```diff
@@ -29,8 +29,8 @@ class Latin1TextIterator {
     // 'endCharacter' denotes the maximum length of the UChar array, which might exceed 'lastIndex'.
-    Latin1TextIterator(const LChar* characters, unsigned currentIndex, unsigned lastIndex, unsigned /*endCharacter*/)
-        : m_characters(characters)
+    Latin1TextIterator(std::span<const LChar> characters, unsigned currentIndex, unsigned lastIndex, unsigned /*endCharacter*/)
```
I don’t understand why this function takes `endCharacter`. Can we remove it?

cdumez replied:
Will check.

darinadler commented on Source/WebCore/platform/graphics/mac/ComplexTextControllerCoreText.mm:
```diff
@@ -177,16 +176,16 @@ static CFDictionaryRef typesetterOptions()
         stream << "Code Units:";
-        for (unsigned i = 0; i < length; ++i)
-            stream << " " << cp[i];
+        for (auto c : cp)
```
Maybe `codePoint` instead of `c`?

cdumez committed:
Drop TextRun::characters8() / characters16()
https://bugs.webkit.org/show_bug.cgi?id=273199

Reviewed by Darin Adler.

Drop TextRun::characters8() / characters16() in favor of span versions.

* Source/WebCore/layout/formattingContexts/inline/text/TextUtil.cpp:
(WebCore::Layout::TextUtil::fallbackFontsForText):
* Source/WebCore/platform/graphics/ComplexTextController.cpp:
(WebCore::ComplexTextController::collectComplexTextRuns):
* Source/WebCore/platform/graphics/FontCascade.cpp:
(WebCore::computeUnderlineType):
* Source/WebCore/platform/graphics/TextRun.h:
* Source/WebCore/platform/graphics/WidthIterator.cpp:
(WebCore::WidthIterator::advance):

Canonical link: https://commits.webkit.org/277986@main
